rhp/v4/server.go

```diff
@@ -0,0 +1,1078 @@
+package rhp
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	"go.sia.tech/core/consensus"
+	rhp4 "go.sia.tech/core/rhp/v4"
+	"go.sia.tech/core/types"
+	"go.sia.tech/coreutils/chain"
+	"go.sia.tech/coreutils/wallet"
+	"go.uber.org/zap"
+	"lukechampine.com/frand"
+)
+
+// maxBasisDiff is the maximum number of blocks by which a transaction's basis
+// can differ from the current tip.
+const maxBasisDiff = 20
+
+var protocolVersion = [3]byte{0, 0, 1}
+
+type (
+	// Usage contains the revenue and risked collateral for a contract.
+	Usage struct {
+		RPCRevenue       types.Currency `json:"rpc"`
+		StorageRevenue   types.Currency `json:"storage"`
+		EgressRevenue    types.Currency `json:"egress"`
+		IngressRevenue   types.Currency `json:"ingress"`
+		AccountFunding   types.Currency `json:"accountFunding"`
+		RiskedCollateral types.Currency `json:"riskedCollateral"`
+	}
+)
+
+type (
+	// A TransportListener is a generic multiplexer for incoming streams.
+	TransportListener interface {
+		AcceptStream() (net.Conn, error)
+		Close() error
+	}
```

Comment:
this makes me wonder if `mux` should implement `net.Listener` 🤔. But I suppose then you have to return a `net.Conn` instead of a `mux.Stream`, which kinda sucks.

Reply:
Anyway I would call this `Mux`